### PR TITLE
Vim.Type: Add is_comparable() function

### PR DESCRIPTION
**Pull request by thinca**
This is useful when user want to compare any two values.

### CODE REVIEW

**Comment by tyru** [comment]
oops... sorry.
I missed the test result seems failing...
But LGTM if the results become success.

**Comment by thinca** [comment]
fmm... `number` and `funcref` can be compared in Vim 8.0, but it can not in Vim 7.4 🤔 
Should I change the behavior in each versions...?

**Comment by tyru** [comment]
> Should I change the behavior in each versions...?

I think so.

Well, sorry to say now, I wonder this function is useful.
Because you can compare two values without errors as follows, how about this?

```vim
try
  return V1 ==# V2
catch
  return 0
endtry
```

File: autoload/vital/__vital__/Vim/Type.vim [hunk 1]
```diff
@@ -47,3 +47,18 @@ function! s:is_predicate(value) abort
...
+  let tname2 = s:type_names[t2]
+  if tname1 ==# 'list' || tname1 ==# 'dict' ||
+  \   tname2 ==# 'list' || tname2 ==# 'dict'
+    return tname1 ==# tname2
+  endif
+  if (tname2 ==# 'job' || tname2 ==# 'channel') &&
+  \   (tname1 ==# 'number' || tname1 ==# 'float')
+    return 0
+  endif
+  return 1
```
Comment (rhysd) [inline autoload/vital/__vital__/Vim/Type.vim]: Possible another implementation:

```vim
function! s:is_comparable(v1, v2) abort
  try
    a:v1 == a:v2
    return 1
  catch
    return 0
  endtry
endfunction
```

**Comment by rhysd** [comment]
ah, sorry for the duplication. My browser shows outdated cache of this page and I did not notice the comment...

**Comment by thinca** [comment]
Hmm... I think `try` should not be used for normal pass.

**Comment by thinca** [comment]
I tried to compare performance between the current implementation and using try-catch by following code.

```
    It performance
      let v = repeat(vals, 30)

      let t = reltime()
      for V1 in v
        for V2 in v
          call T.is_comparable(V1, V2)
        endfor
      endfor
      let t = reltime(t)
      call themis#log(reltimestr(t))

      let t = reltime()
      for V1 in v
        for V2 in v
          call T.is_comparable2(V1, V2)
        endfor
      endfor
      let t = reltime(t)
      call themis#log(reltimestr(t))
    End
```

```
function! s:is_comparable2(value1, value2) abort
  try
    let _ = a:value1 == a:value2
    return 1
  catch
    return 0
  endtry
endfunction
```

As result, using try-catch is faster.

```
  2.678315
  1.655016
```

I imagined `try-catch` to be slow, but it was different.
I'll use `try-catch` here for performance.  Thanks @tyru and @rhysd !

**Comment by thinca** [comment]
I force-pushed and the old implementation was lost 🙇 

I put here the old (called `current` in the previous comment) implementation for the someday.

```vim
function! s:is_comparable(value1, value2) abort
  let [t1, t2] = sort([type(a:value1), type(a:value2)])
  let tname1 = s:type_names[t1]
  let tname2 = s:type_names[t2]
  if tname1 ==# 'list' || tname1 ==# 'dict' ||
  \   tname2 ==# 'list' || tname2 ==# 'dict'
    return tname1 ==# tname2
  endif
  if (tname2 ==# 'job' || tname2 ==# 'channel') &&
  \   (tname1 ==# 'number' || tname1 ==# 'float')
    return 0
  endif
  return 1
endfunction
```

**Comment by thinca** [comment]
All tests passed 🎉

**Comment by k-takata** [comment]
I think `==` can be slow when `value1` and `value2` are both long lists.

**Comment by thinca** [comment]
I agree...

**Comment by k-takata** [comment]
Profiled: https://gist.github.com/k-takata/7a692bceed281fce23e56d4d179c52b7
My expectation was wrong. Comparing long lists was not so slow. However, `[] == {}` was slow. I don't know why.

**Comment by thinca** [comment]
New approach: Cache the results in first time and refernce it later.
This is fastest when function is called many time.
https://wandbox.org/permlink/LXBI7VKSeiQKJaDJ

But, first time make be slow a little.  I think this is acceptable.
https://wandbox.org/permlink/P7iAywfq0hqCpVgf

File: autoload/vital/__vital__/Vim/Type.vim [hunk 2]
```diff
@@ -47,3 +47,37 @@ function! s:is_predicate(value) abort
   return t == s:types.number || t == s:types.string ||
   \ t == s:types.bool || t == s:types.none
 endfunction
+
+function! s:is_comparable(value1, value2) abort
+  if !exists('s:is_comparable_cache')
+    let s:is_comparable_cache = s:_make_is_comparable_cache()
+  endif
+  return s:is_comparable_cache[type(a:value1)][type(a:value2)]
```
Comment (tyru) [inline autoload/vital/__vital__/Vim/Type.vim]: LGTM. But I have one **suggestion**:
* To make cache faster, and less memory footprint, how about constraining `t1 <= t2`?

```vim
let [t1, t2] = [type(a:value1), type(a:value2)]
return s:is_comparable_cache[t1 <= t2 ? t1 : t2][t1 <= t2 ? t2 : t1]
```

and `s:_make_is_comparable_cache()` makes only the necessary results.

Comment (thinca) [reply autoload/vital/__vital__/Vim/Type.vim]: It seems maybe make slow to access cache.
Could you measure performance of making cache and accessing to cache in current code and suggested code?

Comment (tyru) [reply autoload/vital/__vital__/Vim/Type.vim]: sorry.
on second thought, I think it's too early optimization now.
If we do it, let's open a new pull request.

**Pull request merged**